skglm/solvers/pn_solver.py

```diff
@@ -0,0 +1,295 @@
+    for idx, j in enumerate(ws):
+        lipschitz[idx] = raw_hess @ X[:, j] ** 2
+
+    cached_grads = np.zeros(len(ws))
+    X_delta_w_ws = np.zeros(X.shape[0])
+    w_ws = w_epoch[ws]
+
+    for cd_iter in range(max_cd_iter):
+        for idx, j in enumerate(ws):
+
+            # skip when X[:, j] == 0
+            if lipschitz[idx] == 0:
+                continue
+
+            cached_grads[idx] = grad_ws[idx] + X[:, j] @ (raw_hess * X_delta_w_ws)
+            old_w_idx = w_ws[idx]
+            stepsize = 1 / lipschitz[idx]
+
+            w_ws[idx] = penalty.prox_1d(
+                old_w_idx - stepsize * cached_grads[idx],
+                stepsize, j
+            )
+
+            if w_ws[idx] != old_w_idx:
+                X_delta_w_ws += (w_ws[idx] - old_w_idx) * X[:, j]
+
+        if cd_iter % 5 == 0:
+            opt = penalty.subdiff_distance(w_ws, cached_grads, ws)
+            if np.max(opt) <= tol:
+                break
+
+    # descent direction
+    return w_ws - w_epoch[ws], X_delta_w_ws
+
+
+# sparse version of func above
+@njit
+def _compute_descent_direction_s(X_data, X_indptr, X_indices, y,
+                                 w_epoch, Xw_epoch, grad_ws, datafit, penalty,
+                                 ws, max_cd_iter, tol):
+    raw_hess = datafit.raw_hessian(y, Xw_epoch)
+
+    lipschitz = np.zeros(len(ws))
+    for idx, j in enumerate(ws):
+        for i in range(X_indptr[j], X_indptr[j+1]):
+            lipschitz[idx] += raw_hess[X_indices[i]] * X_data[i] ** 2
+
+    cached_grads = np.zeros(len(ws))
+    X_delta_w_ws = np.zeros(len(y))
+    w_ws = w_epoch[ws]
+
+    for cd_iter in range(max_cd_iter):
+        for idx, j in enumerate(ws):
+
+            # skip when X[:, j] == 0
+            if lipschitz[idx] == 0:
+                continue
+
+            cached_grads[idx] = grad_ws[idx]
+            for i in range(X_indptr[j], X_indptr[j+1]):
+                row_i = X_indices[i]
+                cached_grads[idx] += X_data[i] * raw_hess[row_i] * X_delta_w_ws[row_i]
+
+            old_w_idx = w_ws[idx]
+            stepsize = 1 / lipschitz[idx]
+
+            w_ws[idx] = penalty.prox_1d(
+                old_w_idx - stepsize * cached_grads[idx],
+                stepsize, j
+            )
+
+            if w_ws[idx] != old_w_idx:
+                for i in range(X_indptr[j], X_indptr[j+1]):
+                    X_delta_w_ws[X_indices[i]] += (w_ws[idx] - old_w_idx) * X_data[i]
+
+        if cd_iter % 5 == 0:
+            opt = penalty.subdiff_distance(w_ws, cached_grads, ws)
+            if np.max(opt) <= tol:
+                break
+
+    # descent direction
+    return w_ws - w_epoch[ws], X_delta_w_ws
+
+
+@njit
+def _backtrack_line_search(X, y, w, Xw, datafit, penalty, delta_w_ws, X_delta_w_ws,
+                           ws, max_backtrack_iter):
+    # inplace update of w and Xw
+    # return grad_ws of the last w and Xw
+    step, prev_step = 1., 0.
+    prev_penalty_val = penalty.value(w[ws])
```

Comment:
doesn't this need to be updated ?

Reply:
``prev_penalty_val`` refers to the value of the penalty in the epoch loop.
sorry, perhaps the name of the var is a bit misleading. I renamed the variable to avoid confusion.